lib/workload.hpp

```diff
@@ -127,98 +129,146 @@ ostream& operator<<(ostream& os, const rhs_t& rhs);
 bool operator==(const rhs_t& rhs1, const rhs_t& rhs2);
 bool operator<(const rhs_t& rhs1, const rhs_t& rhs2);
 
+//************************************* WlSpec *************************************//
+
+class WlSpec {
+public:
+    virtual ~WlSpec() = default;
+    virtual bool applies_to_queue(unsigned int queue) const = 0;
+    virtual bool equals(const WlSpec& other) const = 0;
+    virtual bool less_than(const WlSpec& other) const = 0;
+    virtual int type_id() const = 0;
+    virtual std::string to_string() const = 0;
+    virtual bool is_available_for_search() const { return false; }
+};
+
+//typedef variant<Comp, Same, Incr, Decr, Unique> wl_spec_t;
+
+bool wl_spec_applies_to_queue(const WlSpec& spec, unsigned int queue);
+bool wl_spec_applies_to_queue(const std::shared_ptr<WlSpec>& spec, unsigned int queue);
+
+bool wl_spec_is_empty(const WlSpec& spec);
+bool wl_spec_is_empty(const std::shared_ptr<WlSpec>& spec);
+
+bool wl_spec_is_all(const WlSpec& spec);
+bool wl_spec_is_all(const std::shared_ptr<WlSpec>& spec);
+
+unsigned int wl_spec_ast_size(const WlSpec& spec);
+unsigned int wl_spec_ast_size(const std::shared_ptr<WlSpec>& wl_spec);
+
+ostream& operator<<(ostream& os, const WlSpec& wl_spec);
+ostream& operator<<(ostream& os, const WlSpec* wl_spec);
+
+bool operator==(const WlSpec& wl_spec1, const WlSpec& wl_spec2);
+bool operator!=(const WlSpec& wl_spec1, const WlSpec& wl_spec2);
+bool operator>=(const WlSpec& wl_spec1, const WlSpec& wl_spec2);
+bool operator>(const WlSpec& wl_spec1, const WlSpec& wl_spec2);
+bool operator<=(const WlSpec& wl_spec1, const WlSpec& wl_spec2);
+bool operator<(const WlSpec& wl_spec1, const WlSpec& wl_spec2);
+
 //************************************* UNIQ *************************************//
-class Unique {
+class Unique : public WlSpec {
 
 public:
     Unique(metric_t metric, qset_t qset);
 
-    bool applies_to_queue(unsigned int queue) const;
+    bool applies_to_queue(unsigned int queue) const override;
 
     qset_t get_qset() const;
     metric_t get_metric() const;
 
+    bool equals(const WlSpec& other) const override;
+    bool less_than(const WlSpec& other) const override;
+    int type_id() const override;
```

Comment:
What is type_id?

Reply:
Can be handled by moving operator inside WlSpec class